AX: Add a failsafe to the while loop in AXCoreObject::partialOrder

twilco opened the pull request:
#### c187483386f14290f1a955ff30821e59af3ee4b1
<pre>
AX: Add a failsafe to the while loop in AXCoreObject::partialOrder
<a href="https://bugs.webkit.org/show_bug.cgi?id=295222">https://bugs.webkit.org/show_bug.cgi?id=295222</a>
<a href="https://rdar.apple.com/154696128">rdar://154696128</a>

Reviewed by Joshua Hoffman.

When testing a different bug, I was able to enter an infinite loop in this function. I tried to rebuild with logging
to debug, but wasn&apos;t able to hit it again despite much effort.

To avoid a permanent hang for our users, add a failsafe counter, and an ASSERT which should hopefully help us debug the
root cause when it triggers.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::partialOrder):

Canonical link: <a href="https://commits.webkit.org/296846@main">https://commits.webkit.org/296846@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd4754d70f8e64c8c9229beb80baad7f07c33396

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109740 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29398 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115761 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59974 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83395 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23963 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98826 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63858 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16971 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59555 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17012 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118553 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36779 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27242 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92398 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37152 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95088 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92220 "Found 2 new API test failures: /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23501 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37187 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14925 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32607 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36674 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42144 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36334 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39676 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1591,11 +1591,19 @@ std::partial_ordering AXCoreObject::partialOrder(const AXCoreObject& other)
+        ++failsafeCounter;
+
         if (RefPtr maybeParent = current ? current->parentObject() : nullptr) {
             if (maybeParent == &other) {
```
I wonder if it'd be worth a debug assert that maybeParent isn't the same as current? While of course that should never happen, that feels like one of the most likely ways we could end up in an endless loop here

twilco replied:
Good idea! Done in latest commit.

minorninth reviewed (commented):
lgtm

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1591,12 +1591,22 @@ std::partial_ordering AXCoreObject::partialOrder(const AXCoreObject& other)
+    // For now, add a failsafe and ASSERT() so we can try to debug the root cause when it does
+    // happen again.
+    unsigned failsafeCounter = 0;
+    static constexpr unsigned maxIterations = Settings::defaultMaximumRenderTreeDepth * 2;
```
Why do we need 2* the maximum depth? A comment explaining that might be helpful here

twilco replied:
To ensure, without a doubt, that this failsafe only triggers when something is wrong. Just `Settings::defaultMaximumRenderTreeDepth` would be wrong — the accessibility tree can be deeper than that when considering things like scrollviews, which in the ancestry chain but not associated with any renderer.

Comment added in latest commit

twilco committed:
AX: Add a failsafe to the while loop in AXCoreObject::partialOrder
https://bugs.webkit.org/show_bug.cgi?id=295222
rdar://154696128

Reviewed by Joshua Hoffman.

When testing a different bug, I was able to enter an infinite loop in this function. I tried to rebuild with logging
to debug, but wasn't able to hit it again despite much effort.

To avoid a permanent hang for our users, add a failsafe counter, and an ASSERT which should hopefully help us debug the
root cause when it triggers.

* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::partialOrder):

Canonical link: https://commits.webkit.org/296846@main
